csp/handler.go

```diff
@@ -161,16 +209,59 @@ func (csp *BlindCSP) sharedKeyReq(msg *bearerstdapi.BearerStandardAPIdata,
 	if len(pid) != processIDSize {
 		return fmt.Errorf("wrong process id: %x", pid)
 	}
-	resp := Message{}
-	var ok bool
-	if ok, resp.Response = csp.AuthCallback(ctx.Request, req, pid, SignatureTypeSharedKey); ok {
-		resp.SharedKey, err = csp.SharedKey(pid)
-		if err != nil {
-			return err
+	// Auth Step
+	step, err := strconv.Atoi(ctx.URLParam("step"))
+	if err != nil {
+		// For backwards compatibility
+		step = 0
+	}
+
+	var resp types.Message
+	var authResp handlers.AuthResponse
+	if authResp = csp.callbacks.Auth(
+		ctx.Request,
+		req,
+		pid,
+		types.SignatureTypeSharedKey,
+		step,
+	); authResp.Success {
+		if authResp.AuthToken == nil {
+			resp.SharedKey, err = csp.SharedKey(pid)
+			if err != nil {
+				return err
+			}
 		}
 	} else {
 		return fmt.Errorf("unauthorized")
 	}
+	resp.Response = authResp.Response
+	resp.AuthToken = authResp.AuthToken
+	return ctx.Send(resp.Marshal(), bearerstdapi.HTTPstatusCodeOK)
+}
+
+func (csp *BlindCSP) info(msg *bearerstdapi.BearerStandardAPIdata,
+	ctx *httprouter.HTTPContext) error {
+	if csp.callbacks.Info == nil {
+		return ctx.Send(nil, bearerstdapi.HTTPstatusCodeOK)
+	}
+	resp := csp.callbacks.Info()
+	if resp == nil {
+		return ctx.Send(nil, bearerstdapi.HTTPstatusCodeOK)
+	}
+	return ctx.Send(resp.Marshal(), bearerstdapi.HTTPstatusCodeOK)
+}
+
+func (csp *BlindCSP) indexer(msg *bearerstdapi.BearerStandardAPIdata,
+	ctx *httprouter.HTTPContext) error {
+	if csp.callbacks.Indexer == nil {
+		return ctx.Send(nil, bearerstdapi.HTTPstatusCodeOK)
+	}
+	userID, err := hex.DecodeString(trimHex(ctx.URLParam("userId")))
+	if err != nil {
+		return fmt.Errorf("cannot get user id: %w", err)
+	}
+	var resp types.Message
+	resp.Elections = csp.callbacks.Indexer(userID)
```

Comment:
I'm getting lost in this code; where is the Indexer method actually implemented?

Reply:
By the handler interface and its implementations. All under handlers/ dir